cpp/src/hdbscan/detail/extract.cuh

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <label/classlabels.cuh>
+
+#include <cub/cub.cuh>
+
+#include <raft/cudart_utils.h>
+
+#include "membership.cuh"
+#include "select.cuh"
+#include "stabilities.cuh"
+#include "utils.h"
+
+#include <raft/sparse/op/sort.h>
+#include <raft/sparse/convert/csr.cuh>
+
+#include <cuml/cluster/hdbscan.hpp>
+
+#include <raft/label/classlabels.cuh>
+
+#include <algorithm>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace ML {
+namespace HDBSCAN {
+namespace detail {
+namespace Extract {
+
+/**
+ * Union-rank data structure with path compression for
+ * labeling data points based on their farthest ancestors
+ * under root.
+ *
+ * For correctness, it is important that all children are
+ * visited before their parents.
+ * @tparam value_idx
+ */
+template <typename value_idx>
+class TreeUnionFind {
+ public:
+  TreeUnionFind(value_idx size_) : size(size_), data(size_ * 2, 0) {
+    for (int i = 0; i < size; i++) {
+      data[i * 2] = i;
+    }
+  }
+
+  void perform_union(value_idx x, value_idx y) {
+    value_idx x_root = find(x);
+    value_idx y_root = find(y);
+
+    if (data[x_root * 2 + 1] < data[y_root * 2 + 1])
+      data[x_root * 2] = y_root;
+    else if (data[x_root * 2 + 1] > data[y_root * 2 + 1])
+      data[y_root * 2] = x_root;
+    else {
+      data[y_root * 2] = x_root;
+      data[x_root * 2 + 1] += 1;
+    }
+  }
+
+  value_idx find(value_idx x) {
+    if (data[x * 2] != x) {
+      data[x * 2] = find(data[x * 2]);
+    }
+
+    return data[x * 2];
+  }
+
+  value_idx *get_data() { return data.data(); }
+
+ private:
+  value_idx size;
+  std::vector<value_idx> data;
+};
+
+template <typename value_idx, typename value_t>
+void do_labelling_on_host(
```

Comment:
Here we build on union find structure for the nodes that are not 'cluster nodes', and use that union find structure to look up the cluster label for the leaves.  We could also do the opposite: using the condensed tree, climb down from each cluster node towards the leaves. Wouldn't it be simpler/shorter to implement that using `Select::perform_bfs`, with a frontier that encodes the cluster label?

Reply:
There are a few (somewhat hidden) assumptions here, such as the reverse topological sort order forcing children to be unioned before their parents, which made this less trivial to parallelize than I originally thought. Though I also agree that if we can find a good parallization strategy then we should try to get this on the device. Given that we have extensively verified this function is returning the correct results, and is not yet a bottleneck even at the scale of 3M data samples (google news word2vec dataset), I would like to add a note to investigate this in our checklist of follow-on items. What do you think?

Reply:
My comment here is not about performance. My point with the alternative solution is to reuse the existing `perform_bfs` instead of introducing the union find structure just for this step. But I might miss some of the assumptions that you mention. It is fine to add this to the follow up items.